semantic-core/src/Data/Core/Pretty.hs

```diff
@@ -56,57 +55,58 @@ inParens amount go = do
   body <- with amount go
   pure (encloseIf (amount >= prec) (symbol "(") (symbol ")") body)
 
-prettify :: (Member (Reader [AnsiDoc]) sig, Member (Reader Prec) sig, Carrier sig m)
-         => Style
-         -> Core (Const (m AnsiDoc)) a
-         -> m AnsiDoc
-prettify style = \case
-  Let a -> pure $ keyword "let" <+> name a
-  Const a :>> Const b -> do
-    prec <- ask @Prec
-    fore <- with 12 a
-    aft  <- with 12 b
-
-    let open  = symbol ("{" <> softline)
-        close = symbol (softline <> "}")
-        separator = ";" <> Pretty.line
-        body = fore <> separator <> aft
-
-    pure . Pretty.align $ encloseIf (12 > prec) open close (Pretty.align body)
-
-  Lam n f -> inParens 11 $ do
-    (x, body) <- bind n f
-    pure (lambda <> x <+> arrow <+> body)
-
-  Frame    -> pure $ primitive "frame"
-  Unit     -> pure $ primitive "unit"
-  Bool b   -> pure $ primitive (if b then "true" else "false")
-  String s -> pure . strlit $ Pretty.viaShow s
-
-  Const f :$ Const x -> inParens 11 $ (<+>) <$> f <*> x
-
-  If (Const con) (Const tru) (Const fal) -> do
-    con' <- "if"   `appending` con
-    tru' <- "then" `appending` tru
-    fal' <- "else" `appending` fal
-    pure $ Pretty.sep [con', tru', fal']
-
-  Load (Const p)   -> "load" `appending` p
-  Edge Lexical (Const n) -> "lexical" `appending` n
-  Edge Import (Const n) -> "import" `appending` n
-  Const item :. Const body   -> inParens 4 $ do
-    f <- item
-    g <- body
-    pure (f <> symbol "." <> g)
-
-  Const lhs := Const rhs -> inParens 3 $ do
-    f <- lhs
-    g <- rhs
-    pure (f <+> symbol "=" <+> g)
-
-  -- Annotations are not pretty-printed, as it lowers the signal/noise ratio too profoundly.
-  Ann _ (Const c) -> c
-  where bind (Ignored x) f = let x' = name x in (,) x' <$> local (x':) (getConst (unScope f))
+prettyCore :: Style -> Term Core User -> AnsiDoc
+prettyCore style = run . runReader @Prec 0 . go (pure . name)
+  where go :: (Member (Reader Prec) sig, Carrier sig m) => (a -> m AnsiDoc) -> Term Core a -> m AnsiDoc
```

Comment:
`go` is a direct-recursive worker function, walking the `Term` via a helper continuation which we extend to deal with `Incr`s efficiently as we walk under binders.